lib/mongoose_push/service/apns/supervisor.ex

```diff
@@ -31,45 +31,104 @@ defmodule MongoosePush.Service.APNS.Supervisor do
+    prod_cert_pools =
       pool_configs
-      |> Enum.filter(fn {_, pool_config} -> pool_config[:mode] == :prod end)
-      |> List.foldl([], &convert_pool_to_sparrow/2)
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :prod and
+          pool_config[:auth][:type] == :certificate_based
+      end)
+      |> Enum.map(&convert_cert_pool_to_sparrow/1)
 
-    [{:dev, dev_pools}, {:prod, prod_pools}]
+    {dev_token_pools, dev_tokens} =
+      pool_configs
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :dev and
+          pool_config[:auth][:type] == :token_based
+      end)
+      |> List.foldl({[], []}, &convert_token_pool_to_sparrow/2)
+
+    {prod_token_pools, prod_tokens} =
+      pool_configs
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :prod and
+          pool_config[:auth][:type] == :token_based
+      end)
+      |> List.foldl({[], []}, &convert_token_pool_to_sparrow/2)
+
+    [
+      {:dev, dev_cert_pools ++ dev_token_pools},
+      {:prod, prod_cert_pools ++ prod_token_pools},
+      {:tokens, dev_tokens ++ prod_tokens}
+    ]
   end
 
-  defp convert_pool_to_sparrow({pool_name, pool_config}, mode_list) do
+  defp convert_cert_pool_to_sparrow({pool_name, pool_config}) do
     auth_type = {:auth_type, :certificate_based}
-    cert = {:cert, pool_config[:cert]}
-    key = {:key, pool_config[:key]}
+    cert = {:cert, pool_config[:auth][:cert]}
+    key = {:key, pool_config[:auth][:key]}
     pool_size = {:worker_num, pool_config[:pool_size]}
 
     endpoint_mode = @default_endpoints[pool_config[:mode]]
     endpoint = {:endpoint, pool_config[:endpoint] || endpoint_mode}
 
-    port_config =
-      case pool_config[:use_2197] do
-        true ->
-          2197
+    port_config = if pool_config[:use_2197], do: 2197, else: nil
+    port = {:port, port_config}
+
+    name = {:pool_name, pool_name}
+
+    [auth_type, cert, key, pool_size, endpoint, port, name]
+    |> Enum.filter(fn {_key, value} -> !is_nil(value) end)
+  end
+
+  defp convert_token_pool_to_sparrow({pool_name, pool_config}, {mode_list, tokens}) do
+    auth_type = {:auth_type, :token_based}
+    pool_size = {:worker_num, pool_config[:pool_size]}
 
-        _ ->
-          nil
-      end
+    endpoint_mode = @default_endpoints[pool_config[:mode]]
+    endpoint = {:endpoint, pool_config[:endpoint] || endpoint_mode}
 
+    port_config = if pool_config[:use_2197], do: 2197, else: nil
     port = {:port, port_config}
 
     name = {:pool_name, pool_name}
 
+    key = pool_config[:auth][:key_id]
+    team = pool_config[:auth][:team_id]
+    p8_file_path = pool_config[:auth][:p8_file_path]
+
+    if is_nil(key) or is_nil(team) or not File.exists?(p8_file_path) do
+      Logger.error(~s"Required authentication elements are missing. Got:
+      key=#{key}, team=#{team}, p8_file=#{p8_file_path}")
+
+      Supervisor.stop(self(), {:error, "Required authentication elements are missing. Got:
```

Comment:
Two issues here: 
1. We should return the error and use `{:stop, reason}` return value from `init/1`
2. reason should never be a string - we need something the we can easely match on. The string is logged, which great, and that should be it.